docs/alloy-models/linear-fee-function/linear-fee.als

```diff
@@ -0,0 +1,313 @@
+pred stutter[f: LinearFeeFunction] {
+  // No change in any of our variables.
+  f.startingFeeRate' = f.startingFeeRate
+  f.endingFeeRate' = f.endingFeeRate
+  f.currentFeeRate' = f.currentFeeRate
+  f.position' = f.position 
+  f.width' = f.width
+  f.deltaFeeRate' = f.deltaFeeRate
+  activeFuncs' = activeFuncs
+}
+
+// Event is used to generate easier to follow traces. These these enums are
+// events that correspond to the main state transition predicates above.
+//
+// See this portion of documentation for more information on this idiom:
+// https://haslab.github.io/formal-software-design/modelling-tips/index.html#an-idiom-to-depict-events.
+enum Event { Stutter, Init, FeeBump }
+
+// stutter is a derived relation that's populated whenever a stutter event
+// happens.
+fun stutter : set Event -> LinearFeeFunction {
+  { e: Stutter, f: LinearFeeFunction | stutter[f] }
+}
+
+// init is a derived relation that's populated whenever an init event happens.
+fun init : Event -> LinearFeeFunction -> Int -> Int -> Int {
+  { e: Init, f: LinearFeeFunction, m: Int, s: Int, c: Int | init[f, m, s, c] }
+}
+
+// fee_bump is a derived relation that's populated whenever a fee_bump event
+// happens.
+fun fee_bump : Event -> LinearFeeFunction {
+  { e: FeeBump, f: LinearFeeFunction | fee_bump[f] }
+}
+
+// events is our final derived relation that stores all the possible event
+// types.
+fun events : set Event {
+  stutter.LinearFeeFunction + init.Int.Int.Int.LinearFeeFunction + fee_bump.LinearFeeFunction
+}
+
+// traces is a fact to ensure that we always get one of the above events in
+// traces we generate.
+fact traces {
+  always some events
+}
+
+// init_traces builds on the fact traces above, and also asserts that there's
+// always at least one init event (to ensure proper initialization).
+fact init_traces {
+  eventually (one init)
+}
+
+// fee_bump_happens ensures that within our trace, there's always at least a
+// single fee bump event.
+fact fee_bump_happens {
+  eventually (some fee_bump)
+}
+
+
+// validate_fee_rates is a fact that ensures that the ending fee rate is always
+// greater than the starting fee rate.
+fact valid_fee_rates {
+  all f: LinearFeeFunction | f.endingFeeRate > f.startingFeeRate
+}
+
+// init_correctness is an assertion that makes sure once an init event happens,
+// then we have the function in our set of active functions.
+assert init_correctness {
+  all f: LinearFeeFunction | {
+    eventually (some init) implies 
+      eventually f in ActiveFeeFunctions.activeFuncs
+  }
+}
+
+// check init_correctness
+
+// init_always_happens is a trivial assertion that ensures that an init
+// eventually happens in a trace.
+assert init_always_happens {
+   eventually (some init)
+}
+
+// check init_always_happens
+
+// init_then_fee_bump asserts that within our trace, we'll always have at least
+// a single fee bump event after an init.
+assert init_then_fee_bump {
+  eventually (some init)
+  eventually (some fee_bump)
+}
```

Comment:
shouldn't the first one imply the second? I think it is reductive because of the `init_always_happens` assertion but I think it's more precise to say that the `fee_bump` can only happen if the `init` happens.

Reply:
Yeah I tried that initially, but that finds a counter example as it's possible to have a `stutter` step at any state transition in between the `init` and `fee_bump`.

If I change to `eventually (some init; some fee_bump)` (`p; q` is shorthand for `p && after q`), if finds this counter example:

![Screenshot 2024-07-31 at 6 28 17 PM](https://github.com/user-attachments/assets/bd24694e-2e42-43ed-b8da-a60f5a2ea75a)


![Screenshot 2024-07-31 at 6 28 25 PM](https://github.com/user-attachments/assets/cb4389e8-5fe5-4571-b9e2-bf223be03830)

Reply:
Actually this works, which I think is the stricter assertion you were looking for: 
```alloy
assert init_then_fee_bump {
  eventually (some init => some fee_bump)
}
```

Using that instead.